From b9be084ba133b112f1885bd8b679e52dbe4ffbda Mon Sep 17 00:00:00 2001
From: Ramanathan Choodamani <rchoodam@qti.qualcomm.com>
Date: Mon, 26 May 2025 18:06:38 -0700
Subject: [PATCH 04/11] QSDK:DP:wifi: Add PPE-DS Support for QCN6432 radio

Add support registering and managing the PPEDS ring interrupts

Patch-dependency:
        325-QSDK-DP-wifi-Add-support-for-PPE-DS-Direct-switch-fe.patch
Patch-work: None

Signed-off-by: Nithyanantham Paramasivam <nithp@qti.qualcomm.com>
Signed-off-by: Ramanathan Choodamani <rchoodam@qti.qualcomm.com>
---
 drivers/net/wireless/ath/ath12k/Makefile |   1 +
 drivers/net/wireless/ath/ath12k/ahb.c    | 133 ++++++++++++++++++++++-
 drivers/net/wireless/ath/ath12k/ppe.h    |   1 +
 3 files changed, 134 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ath/ath12k/Makefile
+++ b/drivers/net/wireless/ath/ath12k/Makefile
@@ -23,6 +23,7 @@ ath12k-y += core.o \
 	    fw.o \
 	    p2p.o \
 	    fse.o \
+	    ahb.o \
 	    vendor.o \
 	    accel_cfg.o
 
--- a/drivers/net/wireless/ath/ath12k/ahb.c
+++ b/drivers/net/wireless/ath/ath12k/ahb.c
@@ -19,7 +19,9 @@
 #include "fw.h"
 #include "pcic.h"
 #include "wifi7/hal.h"
+#include "ppe.h"
 
+#define ATH12K_IRQ_PPE_OFFSET 53
 #define ATH12K_IRQ_CE0_OFFSET 4
 #define ATH12K_MAX_UPDS 1
 #define ATH12K_UPD_IRQ_WRD_LEN  18
@@ -85,6 +87,9 @@ static const char *irq_name[ATH12K_IRQ_N
 	"wbm2host-tx-completions-ring2",
 	"wbm2host-tx-completions-ring1",
 	"tcl2host-status-ring",
+	"reo2ppe",
+	"ppe_wbm_rel",
+	"ppe2tcl"
 };
 
 enum ext_irq_num {
@@ -125,6 +130,9 @@ enum ext_irq_num {
 	wbm2host_tx_completions_ring2,
 	wbm2host_tx_completions_ring1,
 	tcl2host_status_ring,
+	reo2ppe,
+	ppe_wbm_rel,
+	ppe2tcl
 };
 
 static u32 ath12k_ahb_read32(struct ath12k_base *ab, u32 offset)
@@ -798,7 +806,124 @@ static void ath12k_ahb_free_ext_irq(stru
 	}
 }
 
-static const struct ath12k_hif_ops ath12k_ahb_hif_ops = {
+#ifdef CPTCFG_ATH12K_PPE_DS_SUPPORT
+int ath12k_ahb_ppeds_register_interrupts(struct ath12k_base *ab, int type, int vector,
+					 int ring_num)
+{
+	int ret = -EINVAL, irq;
+	struct platform_device *pdev = ab->pdev;
+	int irq_idx;
+
+	if (ab->dp->ppe.ppeds_soc_idx == ATH12K_PPEDS_INVALID_SOC_IDX) {
+		ath12k_err(ab, "invalid soc idx in ppeds IRQ registration\n");
+		goto irq_fail;
+	}
+	if (type == HAL_WBM2SW_RELEASE && ring_num == HAL_WBM2SW_PPEDS_TX_CMPLN_RING_NUM) {
+		irq_idx = ATH12K_IRQ_PPE_OFFSET + 1;
+		irq = platform_get_irq_byname(ab->pdev,
+					      irq_name[irq_idx]);
+		if (irq < 0) {
+			ath12k_err(ab, "ppeds RegIRQ: invalid irq:%d for type:%d\n",
+				   irq, type);
+			return irq;
+		}
+		ab->irq_num[irq_idx] = irq;
+		irq_set_status_flags(irq, IRQ_DISABLE_UNLAZY);
+
+		snprintf(&ab->dp->ppe.ppeds_irq_name[PPEDS_IRQ_PPE_WBM2SW_REL][0],
+			 sizeof(ab->dp->ppe.ppeds_irq_name[PPEDS_IRQ_PPE_WBM2SW_REL]),
+			 "ppe_wbm_rel_%d", ab->dp->ppe.ds_node_id);
+
+		ret = devm_request_irq(&pdev->dev, irq,  ath12k_dp_ppeds_handle_tx_comp,
+				       IRQF_NO_AUTOEN | IRQF_NO_SUSPEND,
+				       ab->dp->ppe.ppeds_irq_name[PPEDS_IRQ_PPE_WBM2SW_REL],
+				       (void *)ab);
+
+		if (ret) {
+			ath12k_err(ab, "ppeds RegIRQ: req_irq fail:%d\n", ret);
+			goto irq_fail;
+		}
+		ab->dp->ppe.ppeds_irq[PPEDS_IRQ_PPE_WBM2SW_REL] = irq;
+
+	} else if (type == HAL_PPE2TCL) {
+		irq_idx = ATH12K_IRQ_PPE_OFFSET + 2;
+		irq = platform_get_irq_byname(ab->pdev,
+					      irq_name[irq_idx]);
+		if (irq < 0) {
+			ath12k_err(ab, "ppeds RegIRQ: invalid irq:%d for type:%d\n",
+				   irq, type);
+			return irq;
+		}
+		ab->irq_num[irq_idx] = irq;
+		irq_set_status_flags(irq, IRQ_DISABLE_UNLAZY);
+
+		snprintf(&ab->dp->ppe.ppeds_irq_name[PPEDS_IRQ_PPE2TCL][0],
+			 sizeof(ab->dp->ppe.ppeds_irq_name[PPEDS_IRQ_PPE2TCL]),
+			 "ppe2tcl_%d",  ab->dp->ppe.ds_node_id);
+
+		ret = devm_request_irq(&pdev->dev, irq,  ath12k_ds_ppe2tcl_irq_handler,
+				       IRQF_NO_AUTOEN | IRQF_NO_SUSPEND,
+				       ab->dp->ppe.ppeds_irq_name[PPEDS_IRQ_PPE2TCL],
+				       (void *)ath12k_dp_get_ppe_ds_ctxt(ab));
+		if (ret) {
+			ath12k_err(ab, "ppeds RegIRQ: req_irq fail:%d\n", ret);
+			goto irq_fail;
+		}
+		ab->dp->ppe.ppeds_irq[PPEDS_IRQ_PPE2TCL] = irq;
+
+	} else if (type == HAL_REO2PPE) {
+		irq_idx = ATH12K_IRQ_PPE_OFFSET;
+		irq = platform_get_irq_byname(ab->pdev,
+					      irq_name[irq_idx]);
+		if (irq < 0) {
+			ath12k_err(ab, "ppeds IRQreg: invalid irq:%d for type:%d\n",
+				   irq, type);
+			return irq;
+		}
+		ab->irq_num[irq_idx] = irq;
+		irq_set_status_flags(irq, IRQ_DISABLE_UNLAZY);
+
+		snprintf(&ab->dp->ppe.ppeds_irq_name[PPEDS_IRQ_REO2PPE][0],
+			 sizeof(ab->dp->ppe.ppeds_irq_name[PPEDS_IRQ_REO2PPE]),
+			 "reo2ppe_%d", ab->dp->ppe.ds_node_id);
+
+		ret = devm_request_irq(&pdev->dev, irq,  ath12k_ds_reo2ppe_irq_handler,
+				       IRQF_SHARED | IRQF_NO_SUSPEND,
+				       ab->dp->ppe.ppeds_irq_name[PPEDS_IRQ_REO2PPE],
+				       (void *)ath12k_dp_get_ppe_ds_ctxt(ab));
+		if (ret) {
+			ath12k_err(ab, "ppeds RegIRQ: req_irq fail:%d\n", ret);
+			goto irq_fail;
+		}
+		ab->dp->ppe.ppeds_irq[PPEDS_IRQ_REO2PPE] = irq;
+		disable_irq_nosync(irq);
+	}
+
+	return 0;
+
+irq_fail:
+	return ret;
+}
+
+void ath12k_ahb_ppeds_irq_disable(struct ath12k_base *ab, enum ppeds_irq_type type)
+{
+	disable_irq_nosync(ab->dp->ppe.ppeds_irq[type]);
+}
+
+void ath12k_ahb_ppeds_irq_enable(struct ath12k_base *ab, enum ppeds_irq_type type)
+{
+	enable_irq(ab->dp->ppe.ppeds_irq[type]);
+}
+
+void ath12k_ahb_ppeds_free_interrupts(struct ath12k_base *ab)
+{
+	disable_irq_nosync(ab->dp->ppe.ppeds_irq[PPEDS_IRQ_PPE2TCL]);
+	disable_irq_nosync(ab->dp->ppe.ppeds_irq[PPEDS_IRQ_REO2PPE]);
+	disable_irq_nosync(ab->dp->ppe.ppeds_irq[PPEDS_IRQ_PPE_WBM2SW_REL]);
+}
+#endif
+
+static struct ath12k_hif_ops ath12k_ahb_hif_ops = {
 	.start = ath12k_ahb_start,
 	.stop = ath12k_ahb_stop,
 	.read32 = ath12k_ahb_read32,
@@ -812,6 +937,12 @@ static const struct ath12k_hif_ops ath12
 	.ce_irq_disable = ath12k_ahb_ce_irqs_disable,
 	.ext_irq_setup = ath12k_ahb_config_ext_irq,
 	.ext_irq_cleanup = ath12k_ahb_free_ext_irq,
+#ifdef CPTCFG_ATH12K_PPE_DS_SUPPORT
+	.ppeds_register_interrupts = ath12k_ahb_ppeds_register_interrupts,
+	.ppeds_free_interrupts = ath12k_ahb_ppeds_free_interrupts,
+	.ppeds_irq_enable = ath12k_ahb_ppeds_irq_enable,
+	.ppeds_irq_disable = ath12k_ahb_ppeds_irq_disable,
+#endif
 };
 
 static const struct ath12k_hif_ops ath12k_ahb_hif_ops_qcn6432 = {
--- a/drivers/net/wireless/ath/ath12k/ppe.h
+++ b/drivers/net/wireless/ath/ath12k/ppe.h
@@ -34,6 +34,7 @@ enum ppeds_irq_type {
 #define ATH12K_INVALID_PPE_VP_NUM -1
 #define ATH12K_INVALID_PPE_VP_TYPE -1
 #define ATH12K_INVALID_VP_PROFILE_IDX	-1
+#define ATH12K_PPEDS_INVALID_SOC_IDX -1
 
 #define ATH12K_DP_RX_FSE_FLOW_METADATA_MASK      0xFFFF
 
--- a/drivers/net/wireless/ath/ath12k/wifi7/hal_qcn9274.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/hal_qcn9274.c
@@ -565,6 +565,9 @@ const struct ath12k_hw_regs ipq5424_regs
 		HAL_IPQ5424_CE_WFSS_REG_BASE,
 	.hal_umac_ce1_dest_reg_base = 0x00203000 -
 		HAL_IPQ5424_CE_WFSS_REG_BASE,
+	.hal_ppe_rel_ring_base = 0x0000046c,
+	.hal_reo2ppe_ring_base = 0x00000938,
+	.hal_tcl_ppe2tcl_ring_base_lsb = 0x00000c58,
 };
 
 const struct ath12k_hw_regs qcn6432_regs = {
@@ -653,6 +656,10 @@ const struct ath12k_hw_regs qcn6432_regs
 	.hal_umac_ce0_dest_reg_base = 0x01B81000,
 	.hal_umac_ce1_src_reg_base =  0x01B82000,
 	.hal_umac_ce1_dest_reg_base =  0x01B83000,
+
+	.hal_ppe_rel_ring_base = 0x0000046c,
+	.hal_reo2ppe_ring_base = 0x00000938,
+	.hal_tcl_ppe2tcl_ring_base_lsb = 0x00000c58
 };
 
 const struct ath12k_hw_hal_params ath12k_wifi7_hw_hal_params_ipq5332 = {
--- a/drivers/net/wireless/ath/ath12k/wifi7/hw.c
+++ b/drivers/net/wireless/ath/ath12k/wifi7/hw.c
@@ -831,7 +831,7 @@ static const struct ath12k_hw_params ath
 		.ftm_responder = true,
 		.credit_flow = false,
 		.is_plink_preferable = true,
-		.ds_support = true,
+		.ds_support = false,
 	},
 	{
 		.name = "ipq5424 hw1.0",
